Fixed #25598 -- Added SCRIPT_NAME prefix to STATIC_URL and MEDIA_URL set to relative paths.

okainov opened the pull request:
Original PRs https://github.com/django/django/pull/10724 and #7000 seem abandoned, so let me try to revive this activity based on #10724.

Fixing issue https://code.djangoproject.com/ticket/25598 


Behavior:

SCRIPT_NAME	|STATIC_URL setting	|STATIC_URL final value
------------ | -------------|----
 /subpath	| http://myhost.com/media/ |	http://myhost.com/media/
 /subpath |	media/ |	 /subpath/media/
None or empty |	media/ |	 /media/
 /subpath |	/media/	|  /media/

To get similar behavior with current `{% url %}` tag we apply `SCRIPT_NAME` only for relative path, including the case when `SCRIPT_NAME` is empty or not set. Thus final `*_URL` value will be always absolute.

So it also improves the logic of relative paths in `*_URL` settings, de-facto transforming them to absolute paths.

felixxm commented on docs/releases/2.2.txt:
```diff
@@ -469,6 +469,14 @@ Miscellaneous
+  already contain ``SCRIPT_NAME`` as a prefix. This may cause backwards incompatibility in
+  certain rare corner cases, i.e., where one of these settings is intentionally
+  configured to be identical to the header-supplied ``SCRIPT_NAME``. In these
+  cases, the ``*_URL`` setting should be set to an empty string.
```
Please remove this release notes.

okainov replied:
done

felixxm commented on tests/settings_tests/tests.py:
```diff
@@ -21,7 +23,6 @@
 })
 @override_settings(ITEMS=['a', 'c', 'e'], ITEMS_OUTER=[1, 2, 3], TEST='override', TEST_OUTER='outer')
 class FullyDecoratedTranTestCase(TransactionTestCase):
-
```
Please revert unrelated changes.

okainov replied:
done

felixxm commented on tests/settings_tests/tests.py:
```diff
@@ -567,3 +571,197 @@ def decorated_function():
+                MEDIA_URL=case.initial_media_url, )
+            self.set_script_name(case.script_name)
+            self.assertEqual(s.STATIC_URL, case.final_static_url)
+            self.assertEqual(s.MEDIA_URL, case.final_media_url)
```
This should use `self.subTest`, I also don't think that we need a `namedtuple` and `self.script_name_test_cases` variable, e.g.
```python
tests = (
    # SCRIPT_NAME ends with no slash, settings start with slashes.
    ('/somesubpath', '/static/', '/somesubpath/static/', '/media/', '/somesubpath/media/'),
    # SCRIPT_NAME ends with no slash, settings start with no slashes.
    ('/somesubpath', 'static/', '/somesubpath/static/', 'media/', '/somesubpath/media/'),
    ...
)
for script_name, initial_static_url, final_static_url, initial_media_url, final_media_url in tests:
    with self.subTest(...):
        ...
```

okainov replied:
done

felixxm commented on django/conf/__init__.py:
```diff
@@ -109,6 +110,29 @@ def configure(self, default_settings=global_settings, **options):
+
+        Useful for when the app is being served at a subpath and manually prefixing
+        that subpath to STATIC_URL and STATIC_ROOT in settings is inconvenient
+        (i.e., a deployment concern). Refs #25598.
```
Please chop `Refs #25598.`.

okainov replied:
done

apollo13 commented on django/conf/__init__.py:
```diff
@@ -109,6 +110,29 @@ def configure(self, default_settings=global_settings, **options):
+        if not prefix or prefix == '/':
+            return value
+        # try to detect if the prefix is already prefixed appropriately.
+        if value.startswith(prefix.rstrip('/')):
```
As said in the previous PR already, I think we should drop this

okainov replied:
I quickly looked thru previous PR comments, but didn't really find solid background reasons. Could you please explain why?

Without this we will for sure break a lot of websites which use hardcoded subfolder already. This tries to mitigate this at least a little bit. If you don't mind explicit breaking of backward compatibility then for sure I can remove this part.

apollo13 replied:
I think the conclusion on the ticket/old PR was that prefixing would only happen for relative paths, which is not something that is currently allowed. Looking through the docs below, I see that this isn't really clear currently. From what I can see this PR currently still prefixes even if the url starts with a /, this is not something we should do.

The final behavior that we agreed on was documented in https://github.com/django/django/pull/10724#issuecomment-445946077 -- why did you revert that behavior again?

okainov replied:
Because I do not agree with not prefixing absolute path. This change does not make sense if it does not prefix absolute path. I'm not even sure that there are use cases where one uses relative `STATIC_URL` - because it will not work for pages more than one level deep. YOu kind of always use `/static` and it should work for both local development (`manage.py runserver` and `localhost:8000/static`) and production environment in subfolder (`myserver.com/mytool/folder/page` and `myserver.com/mytool/static`)

okainov replied:
As discussed, done

apollo13 commented:
From the PR description:
> Prefix is required for absolute paths as well, because app should not be aware of subfolder it's running in, thus it should always use SCRIPT_NAME when it's set (except when *_URL is set to a valid URL, of course).

I do not agree with that. If we just do it for relative paths the docs become rather simple: If you use a relative path it will be prefixed with script_prefix, otherwise it is taken as is.
Your suggested behaviour would be backwards incompatible because it is not longer possible to set STATIC_URL to `/static` while the app is served under `/app`

okainov commented:
>Your suggested behaviour would be backwards incompatible because it is not longer possible to set STATIC_URL to /static while the app is served under /app

If the project right now is running under `/app` there there is no way that `STATIC_URL` is set to `/static`. Because it doesn't work like that, you will have to have really dirty hacks on your server to serve Django project from `myserver.com/app` and static files **from this project** from `myserver.com/static` (I'm not sure which meaning you put into "app" word, I prefer to use "Django project" here to do not mix it with "Django application". Because everything in this PR relates to Django projects, not several Django apps in one project). These users will have STATIC_URL set to either `/app/static` or absolute URL to CDN-like server or, if then have only top-level pages or forgot slash `app/static`. Thus prefixation detection will help to keep their app up and running without any changes.

>EDIT:// It is okay that people will have to remove the existing / from their current settings if they want to get the benefits of script name prefixing.

I do not think this is okay. This should be default behavior. SCRIPT_NAME is quite standard header to you should not require users to use dirty workarounds to get benefits of it. BTW if you run nginx+uWSGI then SCRIPT_NAME is not being forwarded by default thru uWSGI, so if users don't have it configured, nothing will go wrong with their current code. If they have it configured, then it's very expected and we should use it.

apollo13 commented:
> If the project right now is running under `/app` there there is no way that `STATIC_URL` is set to `/static`. 

Sure there is, just set it to `/static` -- Django really doesn't care.

> Because it doesn't work like that, you will have to have really dirty hacks on your server to serve Django project from `myserver.com/app` and static files **from this project** from `myserver.com/static` 

It is really easy, just configure for instance nginx like this:
```
server {

    location /media  {
        alias /path/to/your/mysite/media;  # your Django project's media files - amend as required
    }
    location /static {
        alias /path/to/your/mysite/static; # your Django project's static files - amend as required
    }
    location /app {
        proxy_pass django;
    }
}
```
> > EDIT:// It is okay that people will have to remove the existing / from their current settings if they want to get the benefits of script name prefixing.
> 
> I do not think this is okay. This should be default behavior. SCRIPT_NAME is quite standard header so we should not require users to use strange workarounds to get benefits of it.

Well, if we started with Django now we might make it the default behaviour. That said it is not, and your changes would be backwards incompatible while my suggested approach allows to enable the new behaviour while also preserving existing working configurations which separate project and static files into different "folders" (url-wise).

> BTW if one runs nginx+uWSGI then SCRIPT_NAME is not being forwarded by default thru uWSGI, so if users don't have it configured, nothing will go wrong with their current code. If they have it configured, then it's very expected and we should use it.

Not passing through SCRIPT_NAME is a bug in uWSGI if you ask me.

> Just as a refresher from original issue text, current `{% url %}` tag already uses `SCRIPT_NAME` header, so it's the `{% static %}` tag which is currently inconsistent. This PR solves this inconsistency.

Sure, but as you can see there are multiple ways of solving it :)

apollo13 commented:
Btw, urlpatterns in Django also start without a slash. So I think it would be consistent to expect a relative path to mean "relative to the mountpoint"

okainov commented:
>Sure there is, just set it to /static -- Django really doesn't care.

What I meant, is not that it's impossible, but that you don't want to set it like this. Why? Yest, you could, but then it would be more like separated "CDN-like" server, and it would make perfect sense to put full URL there.

>It is really easy, just configure for instance nginx like this:

I know. But if you have your app in subfolder, then you have ten different tools in ten different subfolders. And you will never ever mix up all statics from ten different projects in one folder. Proper path for static files from "app" is "/app/static".


>That said it is not, and your changes would be backwards incompatible while my suggested approach allows to enable the new behaviour while also preserving existing working configurations which separate project and static files into different "folders" (url-wise).

If you use now `STATIC_URL`  as relative path now then all the links from the pages which are not on top level will break, because you typically have one folder with static for your project, not `my_app/static`, and `my_app/my_folder/static` and `my_app//my_folder/subfolder/static` and so on.

>Btw, urlpatterns in Django also start without a slash. So I think it would be consistent to expect a relative path to mean "relative to the mountpoint"

From the [doc for URLs](https://docs.djangoproject.com/en/2.2/topics/http/urls/). If you add slash, nothing changes. URLs have always a forward slash :) 
And URLs inside `{% url %}` tag are not really urls but names of the URLs, so it's not really applicable there at all.

> 
There’s no need to add a leading slash, because every URL has that. For example, it’s articles, not /articles.

apollo13 commented:
> What I meant, is not that it's impossible, but that you don't want to set it like this. Why? Yest, you could, but then it would be more like separated "CDN-like" server, and it would make perfect sense to put full URL there.

Please don't tell me what I want.

> Proper path for static files from "app" is "/app/static".

That is your assumption.

> If you use now `STATIC_URL` as relative path now then all the links from the pages which are not on top level will break, because you typically have one folder with static for your project, not `my_app/static`, and `my_app/my_folder/static` and `my_app//my_folder/subfolder/static` and so on.

Before or after the PR? Before the PR yes, but we are talking about new behaviour here, so I don't see why that would be a problem.
 
> From the [doc for URLs](https://docs.djangoproject.com/en/2.2/topics/http/urls/). URLs have always no leading slash in the backend. 

Yes because Django always sets script_name to at least a slash, so it would be a bit of a problem (that + consistency with include)

> And currently I'm sure almost everyone uses either absolute path or URL. There is no point in using "relative" `STATIC_URL` in current Django IMO.

Currently for sure because a relative `STATIC_URL` currently simply does not work.

okainov commented:
>Currently for sure because a relative STATIC_URL currently simply does not work.

Hmm, maybe that's the issue then? And probably we shouldn't call current PR solution for this issue. Or what do you mean? Is there separated issue \ enhancement request for that?

Even if we make adding SCRIPT_NAME only for relative path, this will not be a usable solution at the moment, because in test environment relative STATIC_URL will be still "simply not working". So we would force users to have conditional switch for local and production STATIC_URL variable. 

>Before or after the PR? Before the PR yes, but we are talking about new behaviour here, so I don't see why that would be a problem.

Appending SCRIPT_NAME will implicitly lead into having STATIC_URL always being an absolute path. So yes, it will "somehow" work, but this will in the very same way break backward compatibility for those who are using relative path (if there are such people)

apollo13 commented:
> Hmm, maybe that's the issue then? And probably we shouldn't call current PR solution for this issue. Or what do you mean? Is there separated issue \ enhancement request for that?

No it is no "issue" per se, it is just how it works currently. And instead of your PR being backwards incompatible I suggested to make it work by only appending `SCRIPT_NAME` to relative URLs (which users would have to opt in to). [We'd then most likely also adjust the default settings.py to set `static/` instead of `/static/`]

> Even if we make adding SCRIPT_NAME only for relative path, this will not be a usable solution at the moment, because in test environment relative STATIC_URL will be still "simply not working". So we would force users to have conditional switch for local and production STATIC_URL variable.

Sure it would, in test environments SCRIPT_NAME is empty and if you join that with `STATIC_URL` you'd get `/static/` (absolute) again (Granted this would require slight changes in the PR).
 
> Appending SCRIPT_NAME will implicitly lead into having STATIC_URL always being an absolute path. So yes, it will "somehow" work, but this will in the very same way break backward compatibility for those who are using relative path (if there are such people)

I doubt there is __anyone__ using a `STATIC_URL` that is relative currently because it breaks as soon as you are trying to use the admin app or any other app that has more than one level of slashes in the URL.

I guess in the end it boils down on whether we see your PR as bugfix or as new feature:
* In the case of a bugfix we'd keep your suggested behaviour and you'd have to write backwards incompatible change notes that say that using a `STATIC_URL` outside of `SCRIPT_NAME` no longer works (not sure how the other core devs would like that).
* In the case of a feature we enable this by *adding* the ability (as in: this doesn't work currently) to set `STATIC_URL` to a relative path and then prefix with `SCRIPT_NAME`

As you can see the latter is backwards compatible by definition, because it is a new feature not touching old behaviour

okainov commented:
@apollo13 thanks for the detailed explanation. I was not considering this PR as a "bigger" fix for relative paths in `*_URL` settings in general, now it became more clear for me. 

>In the case of a feature we enable this by adding the ability (as in: this doesn't work currently) to set STATIC_URL to a relative path and then prefix with SCRIPT_NAME

With everything from above, now it makes more sense. Updated PR.

> in test environments SCRIPT_NAME is empty and if you join that with STATIC_URL you'd get /static/ (absolute) again (Granted this would require slight changes in the PR).

Updated the logic to fulfill this.

> [We'd then most likely also adjust the default settings.py to set static/ instead of /static/]

This would be great then!

apollo13 commented on django/conf/__init__.py:
```diff
@@ -109,6 +110,28 @@ def configure(self, default_settings=global_settings, **options):
+        Prefix header value SCRIPT_NAME to ``value`` under certain circumstances.
+
+        Useful for when the app is being served at a subpath and manually prefixing
+        that subpath to STATIC_URL and STATIC_ROOT in settings is inconvenient
```
`STATIC_ROOT` should probably be `MEDIA_URL`? `*_ROOT` refers to actual fs paths.

okainov replied:
done

apollo13 commented on docs/ref/settings.txt:
```diff
@@ -2006,6 +2006,13 @@ Example: ``"http://media.example.com/"``
+.. note::
+
+    If :setting:`MEDIA_URL` is a relative path (i.e., not a URL), then the
+    ``SCRIPT_NAME`` header's value (or ``/`` if the header is not set) will be
```
`SCRIPT_NAME` is not a header but rather a wsgi environment value.

okainov replied:
done

apollo13 commented on docs/releases/3.0.txt:
```diff
@@ -473,6 +473,15 @@ Miscellaneous
+  ``SCRIPT_NAME`` header prefixed to their values when the settings attribute
+  is requested. This auto-prefixation occurs only if *both* of the following are
+  true of the initial values: a) they are *not* valid URLs and b) they are pointing
+  to relative path. This may cause backwards incompatibility in certain rare corner
```
The note about backwards incompatibility can be dropped I think because a relative path never worked imo

okainov replied:
done

apollo13 commented on tests/settings_tests/tests.py:
```diff
@@ -567,3 +568,125 @@ def decorated_function():
+             'static/', '/somesubpath/static/',
+             'media/', '/somesubpath/media/'),
+
+            # Category: No existing prefix detection
```
I think this category can be dropped

okainov replied:
done

apollo13 commented:
@felixxm This should be ready for review now.

apollo13 commented on django/conf/__init__.py:
```diff
@@ -109,6 +110,28 @@ def configure(self, default_settings=global_settings, **options):
+            return value
+        from django.urls import get_script_prefix  # circular dependency if top-level import
+        prefix = get_script_prefix()
+        return os.path.join(prefix, value.lstrip('/'))
```
One minor change here, please don't use `os.path.join` but `'/'.join(…)`. `os.path.join` uses `os.sep` which doesn't have to be a slash always.

okainov replied:
Done

okainov commented:
Just rebased

langovoi commented:
Any updates?

felixxm commented:
@okainov @apollo13 Thanks! :+1:  I rebased from master, targeted to Django 3.1, simplified tests and pushed minor edits.

rr- commented:
Forcing `MEDIA_URL` to `/` causes my company's application to break in a significant way. We have to pin the version until we rewrite a major portion around file uploads. It would be good to have this behavior optional.

apollo13 commented:
@rr- I do not see much gain in making this behaviour optional because to the best of my knowledge this is backwards compatible. If this actually breaks anything for you I'd like to know why exactly. Could you open a ticket with more details and put `apollo13` into CC?

Also to be clear: We are not forcing `MEDIA_URL` to `/`, we merely add `SCRIPT_NAME` (or failing that `/`) to the beginning of `MEDIA_URL` if (and only if) `MEDIA_URL` was a relative path in the first place (which doesn't make much sense to begin with).

asottile commented:
here's an example breakage from `sentry` -- admittedly several years too late to be relevant:

https://github.com/getsentry/sentry/blob/56ccfff29f1981b261a7b42c22c271f84312b20e/src/sentry/web/urls.py#L78
